File: test_backend.py
import json
import threading
import time
import websocket_server
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from std_msgs.msg import Float64, String, Float32
from sensor_msgs.msg import NavSatFix

# 存储所有连接的客户端
clients = {}

# ROS2 节点和发布者、订阅者
ros2_node = None
publishers = {}
subscribers = {}

def on_message(client, server, message):
    try:
        data = json.loads(message)
        print(f'Received from client: {data}')

        # 根据消息类型进行处理
        if data['type'] == 'publish':
            publish_ros2_message(data['topic'], data['data'])
        elif data['type'] == 'subscribe':
            subscribe_to_ros2_topic(data['topic'])
    except Exception as e:
        print(f'Error processing message: {e}')

def on_client_connect(client, server):
    client_id = client['id']
    clients[client_id] = client
    print(f'Client connected: {client_id}')

def on_client_disconnect(client, server):
    client_id = client['id']
    del clients[client_id]
    print(f'Client disconnected: {client_id}')

def broadcast_message(topic, data):
    message = json.dumps({
        'topic': topic,
        'data': data,
        'timestamp': time.strftime('%Y-%m-%dT%H:%M:%S.%fZ', time.gmtime())
    })
    for client in clients.values():
        if 'connected' in client and client['connected']:
            client['handler'].send(message)

def publish_ros2_message(topic, data):
    publisher = publishers.get(topic)
    if publisher:
        msg = None
        if topic == '/cmd_vel':
            msg = Twist()
            msg.linear.x = data['linear']
            msg.angular.z = data['angular']
        elif topic == '/wheel_left/target':
            msg = Float64()
            msg.data = data
        elif topic == '/wheel_right/target':
            msg = Float64()
            msg.data = data
        elif topic == '/vehicle_params':
            msg = String()
            msg.data = json.dumps({
                'type': data['type'],
                'params': {
                    'wheelRadius': data['wheelRadius'],
                    'vehicleWidth': data['vehicleWidth'],
                    'vehicleLength': data['vehicleLength']
                }
            })
        elif topic == '/pid_params':
            msg = String()
            msg.data = json.dumps({
                'p': data['p'],
                'i': data['i'],
                'd': data['d']
            })
        elif topic == '/pid_params_left':
            msg = String()
            msg.data = json.dumps({
                'p': data['p'],
                'i': data['i'],
                'd': data['d'],
                'wheel': 'left'
            })
        elif topic == '/pid_params_right':
            msg = String()
            msg.data = json.dumps({
                'p': data['p'],
                'i': data['i'],
                'd': data['d'],
                'wheel': 'right'
            })

        if msg:
            publisher.publish(msg)
            print(f'Published to {topic}: {msg}')
    else:
        print(f'Publisher not found for topic: {topic}')

def subscribe_to_ros2_topic(topic):
    # 订阅者已经在 init_ros2 中创建完毕, 不需要再次创建
    print(f'Subscribed to topic: {topic}')

def init_ros2():
    global ros2_node, publishers, subscribers

    rclpy.init()
    ros2_node = Node('dashboard_node')
    print('ROS2 node created')

    # 创建发布者
    publishers = {
        '/cmd_vel': ros2_node.create_publisher(Twist, '/cmd_vel', 10),
        '/wheel_left/target': ros2_node.create_publisher(Float64, '/wheel_left/target', 10),
        '/wheel_right/target': ros2_node.create_publisher(Float64, '/wheel_right/target', 10),
        '/vehicle_params': ros2_node.create_publisher(String, '/vehicle_params', 10),
        '/pid_params': ros2_node.create_publisher(String, '/pid_params', 10),
        '/pid_params_left': ros2_node.create_publisher(String, '/pid_params_left', 10),
        '/pid_params_right': ros2_node.create_publisher(String, '/pid_params_right', 10)
    }

    # 创建订阅者
    subscribers = {
        '/wheel_left/feedback': ros2_node.create_subscription(
            Float64, '/wheel_left/feedback', lambda msg: on_ros2_message('/wheel_left/feedback', msg.data), 10
        ),
        '/wheel_right/feedback': ros2_node.create_subscription(
            Float64, '/wheel_right/feedback', lambda msg: on_ros2_message('/wheel_right/feedback', msg.data), 10
        ),
        '/wheel_left/target_speed': ros2_node.create_subscription(
            Float64, '/wheel_left/target_speed', lambda msg: on_ros2_message('/wheel_left/target_speed', msg.data), 10
        ),
        '/wheel_right/target_speed': ros2_node.create_subscription(
            Float64, '/wheel_right/target_speed', lambda msg: on_ros2_message('/wheel_right/target_speed', msg.data), 10
        ),
        '/vehicle_params_feedback': ros2_node.create_subscription(
            String, '/vehicle_params_feedback', lambda msg: on_ros2_message('/vehicle_params_feedback', json.loads(msg.data)), 10
        ),
        '/gps/fix': ros2_node.create_subscription(
            NavSatFix, '/gps/fix', lambda msg: on_ros2_message('/gps/fix', {
                'latitude': msg.latitude,
                'longitude': msg.longitude,
                'altitude': msg.altitude
            }), 10
        ),
        '/heading': ros2_node.create_subscription(
            Float32, '/heading', lambda msg: on_ros2_message('/heading', msg.data), 10
        ),
        '/pid_params_feedback': ros2_node.create_subscription(
            String, '/pid_params_feedback', lambda msg: on_ros2_message('/pid_params_feedback', json.loads(msg.data)), 10
        ),
        '/pid_params_left_feedback': ros2_node.create_subscription(
            String, '/pid_params_left_feedback', lambda msg: on_ros2_message('/pid_params_left_feedback', json.loads(msg.data)), 10
        ),
        '/pid_params_right_feedback': ros2_node.create_subscription(
            String, '/pid_params_right_feedback', lambda msg: on_ros2_message('/pid_params_right_feedback', json.loads(msg.data)), 10
        )
    }

    print('Subscribers created:')
    for topic, subscriber in subscribers.items():
        print(f'- {topic}')

    spin_thread = threading.Thread(target=rclpy.spin, args=(ros2_node,))
    spin_thread.start()
    print('ROS2 node is spinning')

def on_ros2_message(topic, data):
    print(f'Received ROS2 message on topic {topic}: {data}')
    broadcast_message(topic, data)

def broadcast_message(topic, data):
    message = json.dumps({
        'topic': topic,
        'data': data,
        'timestamp': time.strftime('%Y-%m-%dT%H:%M:%S.%fZ', time.gmtime())
    })
    for client_id, client in clients.items():
        if 'connected' in client and client['connected']:
            try:
                client['handler'].send(message)
            except Exception as e:
                print(f'Error sending message to client {client_id}: {e}')
                del clients[client_id]
                print(f'Client {client_id} disconnected')

def start_server(host='0.0.0.0', port=3001):
    server = websocket_server.WebsocketServer(host, port)
    server.set_fn_new_client(on_client_connect)
    server.set_fn_client_left(on_client_disconnect)
    server.set_fn_message_received(on_message)

    print(f'WebSocket server started, listening on {host}:{port}')
    server.run_forever()

if __name__ == '__main__':
    try:
        init_ros2()
    except Exception as e:
        print(f'Failed to initialize ROS2: {e}')

    start_server()

File: test_backend.js
// test_backend.js
import express from 'express';
import { createServer } from 'http';
import { WebSocketServer, WebSocket } from 'ws';
import rclnodejs from 'rclnodejs';

const app = express();
const server = createServer(app);
const wss = new WebSocketServer({ server });

// 存储 ROS2 节点和发布者的全局引用
let ros2Node = null;
let publishers = {};

// 初始化 ROS2
async function initROS2() {
  try {
    console.log('Initializing RCLNode...');
    await rclnodejs.init();
    console.log('RCLNode initialized');

    ros2Node = new rclnodejs.Node('dashboard_node');
    console.log('Node created');

    // 创建发布者
    publishers = {
      '/cmd_vel': ros2Node.createPublisher('geometry_msgs/msg/Twist', '/cmd_vel'),
      '/wheel_left/target': ros2Node.createPublisher('std_msgs/msg/Float64', '/wheel_left/target'),
      '/wheel_right/target': ros2Node.createPublisher('std_msgs/msg/Float64', '/wheel_right/target'),
      '/vehicle_params': ros2Node.createPublisher('std_msgs/msg/String', '/vehicle_params'),
      // 添加 PID 参数发布者
      '/pid_params': ros2Node.createPublisher('std_msgs/msg/String', '/pid_params'),
      '/pid_params_left': ros2Node.createPublisher('std_msgs/msg/String', '/pid_params_left'),
      '/pid_params_right': ros2Node.createPublisher('std_msgs/msg/String', '/pid_params_right')
    };

    // 创建订阅者
    ros2Node.createSubscription(
      'std_msgs/msg/Float64',
      '/wheel_left/feedback',
      (msg) => {
        broadcastMessage('/wheel_left/feedback', msg.data);
      }
    );

    ros2Node.createSubscription(
      'std_msgs/msg/Float64',
      '/wheel_right/feedback',
      (msg) => {
        broadcastMessage('/wheel_right/feedback', msg.data);
      }
    );

    ros2Node.createSubscription(
      'std_msgs/msg/Float64',
      '/wheel_left/target_speed',
      (msg) => {
        broadcastMessage('/wheel_left/target_speed', msg.data);
      }
    );

    ros2Node.createSubscription(
      'std_msgs/msg/Float64',
      '/wheel_right/target_speed',
      (msg) => {
        broadcastMessage('/wheel_right/target_speed', msg.data);
      }
    );

    ros2Node.createSubscription(
      'std_msgs/msg/String',
      '/vehicle_params_feedback',
      (msg) => {
        try {
          const params = JSON.parse(msg.data);
          broadcastMessage('/vehicle_params_feedback', params);
        } catch (error) {
          console.error('Error parsing vehicle params feedback:', error);
        }
      }
    );

    // 在后端的 initROS2 函数中添加新的订阅者
    ros2Node.createSubscription(
      'sensor_msgs/msg/NavSatFix',  // GPS消息类型
      '/gps/fix',                   // GPS话题
      (msg) => {
        broadcastMessage('/gps/fix', {
          latitude: msg.latitude,
          longitude: msg.longitude,
          altitude: msg.altitude
        });
      }
    );

    ros2Node.createSubscription(
      'std_msgs/msg/Float32',      // 航向角消息类型
      '/heading',                  // 航向角话题
      (msg) => {
        broadcastMessage('/heading', msg.data);
      }
    );

    // 添加 PID 参数反馈订阅
    ros2Node.createSubscription(
      'std_msgs/msg/String',
      '/pid_params_feedback',
      (msg) => {
        try {
          const pidParams = JSON.parse(msg.data);
          broadcastMessage('/pid_params_feedback', pidParams);
        } catch (error) {
          console.error('Error parsing PID params feedback:', error);
        }
      }
    );

    // 可以分别订阅左右轮的 PID 参数反馈
    ros2Node.createSubscription(
      'std_msgs/msg/String',
      '/pid_params_left_feedback',
      (msg) => {
        try {
          const pidParams = JSON.parse(msg.data);
          broadcastMessage('/pid_params_left_feedback', pidParams);
        } catch (error) {
          console.error('Error parsing left wheel PID params feedback:', error);
        }
      }
    );

    ros2Node.createSubscription(
      'std_msgs/msg/String',
      '/pid_params_right_feedback',
      (msg) => {
        try {
          const pidParams = JSON.parse(msg.data);
          broadcastMessage('/pid_params_right_feedback', pidParams);
        } catch (error) {
          console.error('Error parsing right wheel PID params feedback:', error);
        }
      }
    );

    ros2Node.spin();
    console.log('Node is spinning');
    return ros2Node;
  } catch (error) {
    console.error('ROS2 initialization error:', error);
    throw error;
  }
}

// 广播消息给所有客户端
function broadcastMessage(topic, data) {
  wss.clients.forEach(client => {
    if (client.readyState === WebSocket.OPEN) {
      client.send(JSON.stringify({
        topic: topic,
        data: data,
        timestamp: new Date().toISOString()
      }));
    }
  });
}

// WebSocket 连接处理
wss.on('connection', (ws) => {
  console.log('Client connected');
  
  // 处理来自前端的消息
  ws.on('message', (message) => {
    try {
      const data = JSON.parse(message);
      console.log('Received from client:', data);

      // 处理不同类型的消息
      if (data.type === 'publish') {
        const publisher = publishers[data.topic];
        if (publisher) {
          let msg;
          // 根据话题类型创建相应的消息
          switch(data.topic) {
            case '/cmd_vel':
              msg = {
                linear: { x: data.data.linear, y: 0.0, z: 0.0 },
                angular: { x: 0.0, y: 0.0, z: data.data.angular }
              };
              break;
            case '/vehicle_params':
              msg = {
                data: JSON.stringify({
                  type: data.data.type,
                  params: {
                    wheelRadius: data.data.wheelRadius,
                    vehicleWidth: data.data.vehicleWidth,
                    vehicleLength: data.data.vehicleLength
                  }
                })
              };
              break;
            case '/pid_params':
              msg = {
                data: JSON.stringify({
                  p: data.data.p,
                  i: data.data.i,
                  d: data.data.d
                })
              };
              break;
            case '/pid_params_left':
              msg = {
                data: JSON.stringify({
                  p: data.data.p,
                  i: data.data.i,
                  d: data.data.d,
                  wheel: 'left'
                })
              };
              break;
            case '/pid_params_right':
              msg = {
                data: JSON.stringify({
                  p: data.data.p,
                  i: data.data.i,
                  d: data.data.d,
                  wheel: 'right'
                })
              };
              break;
            default:
              msg = { data: data.data };
              break;
          }

          // 发布消息
          publisher.publish(msg);
          console.log('Published to', data.topic, ':', msg);
        } else {
          console.error('Publisher not found for topic:', data.topic);
        }
      }
    } catch (error) {
      console.error('Error processing message:', error);
    }
  });
  
  ws.on('close', () => {
    console.log('Client disconnected');
  });
});

// 启动服务器
const PORT = 3001;
server.listen(PORT, async () => {
  console.log(`Server listening on port ${PORT}`);
  try {
    await initROS2();
    console.log('ROS2 initialization completed');
  } catch (error) {
    console.error('Failed to initialize ROS2:', error);
  }
});

// 优雅关闭
process.on('SIGINT', () => {
  console.log('Shutting down...');
  if (ros2Node) {
    ros2Node.destroy();
  }
  server.close(() => {
    process.exit(0);
  });
});